apps/total-typescript/src/server/routers/progress.ts

```diff
@@ -0,0 +1,40 @@
+import {createRouter} from 'server/createRouter'
+import {z} from 'zod'
+import {getSdk} from '@skillrecordings/database'
+import {SubscriberSchema} from 'lib/convertkit'
+
+export const progressRouter = createRouter().mutation('add', {
+  input: z.object({
+    lessonSlug: z.string(),
+  }),
+  async resolve({ctx, input}) {
+    const {findOrCreateUser, completeLessonProgressForUser} = getSdk()
+    try {
+      const subscriberCookie = ctx.req.cookies['ck_subscriber']
+
+      if (!subscriberCookie) {
+        console.debug('no subscriber cookie')
+        return {error: 'no subscriber found'}
+      }
+
+      const subscriber = SubscriberSchema.parse(JSON.parse(subscriberCookie))
+
+      if (!subscriber) {
+        console.debug('no subscriber cookie')
+        return {error: 'no subscriber found'}
+      }
+
+      const {user} = await findOrCreateUser(subscriber.email_address)
+
+      return await completeLessonProgressForUser({
+        userId: user.id,
+        lessonSlug: input.lessonSlug,
+      })
+    } catch (error) {
+      console.error(error)
+      let message = 'Unknown Error'
+      if (error instanceof Error) message = error.message
+      return {error: message}
+    }
+  },
```

Comment:
this is a rip of the logic/service in the api endpoint that got removed

I'd definitely extract it into a literal service to pull it out of trpc context for testing. [The example has jest and playwright setup](https://github.com/trpc/examples-next-prisma-starter-websockets) so that's next phase